### ISSUE TITLE
[STORM-1934] Race condition between sync-supervisor and sync-processes raises several strange issues

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5716]
<p>There're some strange issues including <a href="https://issues.apache.org/jira/browse/STORM-1933" title="Intermittent test failure on test-multiple-active-storms-multiple-supervisors for supervisor-test " class="issue-link" data-issue-key="STORM-1933"><del>STORM-1933</del></a> and others (which I will file an issue soon) which are related to race condition in supervisor.</p>

<p>As I mentioned to <a href="https://issues.apache.org/jira/browse/STORM-1933" title="Intermittent test failure on test-multiple-active-storms-multiple-supervisors for supervisor-test " class="issue-link" data-issue-key="STORM-1933"><del>STORM-1933</del></a>, basically sync-supervisor relies on zk assignment, and sync-processes relies on local assignment and local workers directory, but in fact sync-supervisor also access local state and take some actions which affects sync-processes. And also Satish left the comment to <a href="https://issues.apache.org/jira/browse/STORM-1933" title="Intermittent test failure on test-multiple-active-storms-multiple-supervisors for supervisor-test " class="issue-link" data-issue-key="STORM-1933"><del>STORM-1933</del></a> describing other issue related to race condition and idea to fix this which is same page on me.</p>

---
<details><summary><i>Originally reported by <a title="Jungtaek Lim" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1934" target="_blank">Race condition between sync-supervisor and sync-processes raises several strange issues</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jungtaek Lim" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-07-10T21:08:23+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user HeartSaVioR opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1934" title="Race condition between sync-supervisor and sync-processes raises several strange issues" class="issue-link" data-issue-key="STORM-1934"><del>STORM-1934</del></a> Fix race condition between sync-supervisor and sync-processes</p>

<ul>
	<li>sync-supervisor just downloads new topology code and writes new local assignment</li>
	<li>shutting down workers and removing topology code is moved to sync-processes</li>
	<li>sync-processes does all of jobs based on local assignment and allocated workers</li>
	<li>remove unused / unneeded codes</li>
</ul>


<p>    Here's my test result for this patch:</p>

<ul>
	<li>`mvn clean install` 5 times: not met supervisor intermittent failure (<a href="https://issues.apache.org/jira/browse/STORM-1933" title="Intermittent test failure on test-multiple-active-storms-multiple-supervisors for supervisor-test " class="issue-link" data-issue-key="STORM-1933"><del>STORM-1933</del></a>)</li>
	<li>will try more times</li>
	<li>kill worker via `kill`, `kill -9`, `restart worker` from UI: no issue on restarting worker</li>
	<li>rebalance topology to change workers (2 -> 3): to test that new assignment has same worker port but different executors compared to assigned workers</li>
	<li>worker is recognized as :disallowed, and killed & relaunched</li>
</ul>


<p>    Rebalance test in details:</p>

<ul class="alternate" type="square">
	<li>Writing new assignment<br/>
    ```<br/>
    6701 {:storm-id "test-topology2-4-1467185073", :executors (<span class="error">&#91;7 7&#93;</span> <span class="error">&#91;5 5&#93;</span> <span class="error">&#91;3 3&#93;</span> <span class="error">&#91;1 1&#93;</span>), :resources <span class="error">&#91;0.0 0.0 0.0&#93;</span>}, <br/>
    6702 {:storm-id "test-topology2-4-1467185073", :executors (<span class="error">&#91;6 6&#93;</span> <span class="error">&#91;4 4&#93;</span> <span class="error">&#91;2 2&#93;</span>), :resources <span class="error">&#91;0.0 0.0 0.0&#93;</span>}<br/>
    ```</li>
</ul>


<ul class="alternate" type="square">
	<li>Assigned executors:<br/>
    ```<br/>
    6701 {:storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;7 7&#93;</span> <span class="error">&#91;5 5&#93;</span> <span class="error">&#91;3 3&#93;</span> <span class="error">&#91;1 1&#93;</span>], :resources #object<span class="error">&#91;org.apache.storm.generated.WorkerResources 0x40c4d31c "WorkerResources(mem_on_heap:0.0, mem_off_heap:0.0, cpu:0.0)"&#93;</span>}, <br/>
    6702 {:storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;6 6&#93;</span> <span class="error">&#91;4 4&#93;</span> <span class="error">&#91;2 2&#93;</span>], :resources #object<span class="error">&#91;org.apache.storm.generated.WorkerResources 0x4ba861f4 "WorkerResources(mem_on_heap:0.0, mem_off_heap:0.0, cpu:0.0)"&#93;</span>}}<br/>
    ```</li>
</ul>


<ul class="alternate" type="square">
	<li>Allocated:<br/>
    ```<br/>
    "2e9bea10-02b7-4e55-88e7-b194b9917a63" [:disallowed {:time-secs 1467185407, :storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;3 3&#93;</span> <span class="error">&#91;6 6&#93;</span> <span class="error">&#91;-1 -1&#93;</span>], :port 6703}], <br/>
    "4630c4bf-9786-47ff-9f3b-6b42d9781b9d" [:disallowed {:time-secs 1467185407, :storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;7 7&#93;</span> <span class="error">&#91;1 1&#93;</span> <span class="error">&#91;-1 -1&#93;</span> <span class="error">&#91;4 4&#93;</span>], :port 6701}], <br/>
    "b9a622d2-5e5b-4311-999c-8c8dd92da6b6" [:disallowed {:time-secs 1467185406, :storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;-1 -1&#93;</span> <span class="error">&#91;5 5&#93;</span>], :port 6702}]}<br/>
    ```</li>
</ul>


<p>    NOTE: Due to forward reference, I have to move `sync-processes` to just before `mk-synchronize-supervisor`. Major changes are done in sync-processes so reviewers need to compare before & after manually. Sorry about that.</p>

<p>    Since supervisor.clj is already ported to Java in master branch, I should have time to read ported code, and modify to be in sync.</p>

<p>    Please review and comment while I'm working against master branch. Thanks!</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/HeartSaVioR/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/HeartSaVioR/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-1934" title="Race condition between sync-supervisor and sync-processes raises several strange issues" class="issue-link" data-issue-key="STORM-1934"><del>STORM-1934</del></a>-1.x</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1528</p>

<hr />
<p>commit e5857e07838af888988691af39efbe415b9a2345<br/>
Author: Jungtaek Lim <kabhwan@gmail.com><br/>
Date:   2016-06-29T07:06:20Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1934" title="Race condition between sync-supervisor and sync-processes raises several strange issues" class="issue-link" data-issue-key="STORM-1934"><del>STORM-1934</del></a> Fix race condition between sync-supervisor and sync-processes</p>

<ul>
	<li>sync-supervisor just downloads new topology code and writes new local assignment</li>
	<li>shutting down workers and removing topology code is moved to sync-processes</li>
	<li>sync-processes does all of jobs based on local assignment and allocated workers</li>
	<li>remove unused / unneeded codes</li>
</ul>


<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    Note: This patch removes the change from <a href="https://issues.apache.org/jira/browse/STORM-1561" title="Supervisor should relaunch worker if assignments have changed" class="issue-link" data-issue-key="STORM-1561"><del>STORM-1561</del></a> since supervisor originally checks the change of executors in same port. <br/>
    Please refer `read-allocated-workers` and `matches-an-assignment?`, and let me know if I'm missing here.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Linking other suspected issues.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    Build passes on storm-core so it's unrelated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user satishd commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528#discussion_r68914596" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528#discussion_r68914596</a></p>

<p>    &#8212; Diff: storm-core/src/clj/org/apache/storm/daemon/supervisor.clj &#8212;<br/>
    @@ -592,24 +577,13 @@<br/>
   (doseq [p (set/difference (set (keys existing-assignment))<br/>
     (set (keys new-assignment)))]<br/>
     (.killedWorker isupervisor (int p)))<br/>
    &#8212; End diff &#8211;</p>

<p>     `StandaloneSupervisor` implementation does nothing. What is the need of this API?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user satishd commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528#discussion_r68914602" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528#discussion_r68914602</a></p>

<p>    &#8212; Diff: storm-core/src/clj/org/apache/storm/daemon/supervisor.clj &#8212;<br/>
    @@ -592,24 +577,13 @@<br/>
   (doseq [p (set/difference (set (keys existing-assignment))<br/>
     (set (keys new-assignment)))]<br/>
     (.killedWorker isupervisor (int p)))</p>
<ul class="alternate" type="square">
	<li>(kill-existing-workers-with-change-in-components supervisor existing-assignment new-assignment)<br/>
   (.assigned isupervisor (keys new-assignment))
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>     `StandaloneSupervisor` implementation does nothing. What is the need of this API?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user HeartSaVioR opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1529" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1529</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1934" title="Race condition between sync-supervisor and sync-processes raises several strange issues" class="issue-link" data-issue-key="STORM-1934"><del>STORM-1934</del></a> Fix race condition between sync-supervisor and sync-processes</p>

<ul>
	<li>sync-supervisor just downloads new topology code and writes new local assignment</li>
	<li>shutting down workers and removing topology code is moved to sync-processes</li>
	<li>sync-processes does all of jobs based on local assignment and allocated workers</li>
</ul>


<p>    PR for 1.x-branch: #1528</p>

<p>    Please note that supervisor is ported to Java, so PR for master and PR for 1.x-branch are having different diff. But concept is same.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/HeartSaVioR/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/HeartSaVioR/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-1934" title="Race condition between sync-supervisor and sync-processes raises several strange issues" class="issue-link" data-issue-key="STORM-1934"><del>STORM-1934</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1529.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1529.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1529</p>

<hr />
<p>commit 816a6cc47e5be505d4857e25a4aa14d2f09cade0<br/>
Author: Jungtaek Lim <kabhwan@gmail.com><br/>
Date:   2016-06-29T10:06:43Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1934" title="Race condition between sync-supervisor and sync-processes raises several strange issues" class="issue-link" data-issue-key="STORM-1934"><del>STORM-1934</del></a> Fix race condition between sync-supervisor and sync-processes</p>

<ul>
	<li>sync-supervisor just downloads new topology code and writes new local assignment</li>
	<li>shutting down workers and removing topology code is moved to sync-processes</li>
	<li>sync-processes does all of jobs based on local assignment and allocated workers</li>
</ul>


<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @HeartSaVioR The changes looks good. May be you can add some comments in sync-process and mk-synchronize-supervisor on what each is supposed to do.</p>

<p>    In your comment where you rebalanced from 2 -> 3 workers, the allocated status for the new worker ":port 6703" is also disallowed, which looks odd since its a new worker.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @arunmahadevan Sorry for confusing: rebalance is done from 3 workers to 2 workers, and since all executors map are changed so all workers are disallowed and two new workers are launched.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @HeartSaVioR thanks for the clarification, it makes sense now. Can you also check rebalance from 2 -> 3 workers ?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @arunmahadevan <br/>
    This works perfectly.</p>

<ul class="alternate" type="square">
	<li>Writing new assignment<br/>
    ```<br/>
    6700 {:storm-id "test-topology2-4-1467185073", :executors (<span class="error">&#91;7 7&#93;</span> <span class="error">&#91;4 4&#93;</span> <span class="error">&#91;1 1&#93;</span>), :resources <span class="error">&#91;0.0 0.0 0.0&#93;</span>}, <br/>
    6702 {:storm-id "test-topology2-4-1467185073", :executors (<span class="error">&#91;6 6&#93;</span> <span class="error">&#91;3 3&#93;</span>), :resources <span class="error">&#91;0.0 0.0 0.0&#93;</span>}, <br/>
    6701 {:storm-id "test-topology2-4-1467185073", :executors (<span class="error">&#91;5 5&#93;</span> <span class="error">&#91;2 2&#93;</span>), :resources <span class="error">&#91;0.0 0.0 0.0&#93;</span>}<br/>
    ```</li>
</ul>


<ul class="alternate" type="square">
	<li>Assigned executors:<br/>
    ```<br/>
    {6700 {:storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;7 7&#93;</span> <span class="error">&#91;4 4&#93;</span> <span class="error">&#91;1 1&#93;</span>], :resources #object<span class="error">&#91;org.apache.storm.generated.WorkerResources 0x6dabc02d "WorkerResources(mem_on_heap:0.0, mem_off_heap:0.0, cpu:0.0)"&#93;</span>}, <br/>
    6701 {:storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;5 5&#93;</span> <span class="error">&#91;2 2&#93;</span>], :resources #object<span class="error">&#91;org.apache.storm.generated.WorkerResources 0x6de46954 "WorkerResources(mem_on_heap:0.0, mem_off_heap:0.0, cpu:0.0)"&#93;</span>}, <br/>
    6702 {:storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;6 6&#93;</span> <span class="error">&#91;3 3&#93;</span>], :resources #object<span class="error">&#91;org.apache.storm.generated.WorkerResources 0x60f1fbb5 "WorkerResources(mem_on_heap:0.0, mem_off_heap:0.0, cpu:0.0)"&#93;</span>}}<br/>
    ```</li>
</ul>


<ul class="alternate" type="square">
	<li>Allocated:<br/>
    ```<br/>
    "612eeee0-0ddc-4820-9062-c79276d56cf2" [:disallowed {:time-secs 1467209575, :storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;6 6&#93;</span> <span class="error">&#91;-1 -1&#93;</span> <span class="error">&#91;4 4&#93;</span>], :port 6702}], <br/>
    "9751163a-f7b8-46a5-9b9c-bcd1fd1446cd" [:disallowed {:time-secs 1467209574, :storm-id "test-topology2-4-1467185073", :executors [<span class="error">&#91;7 7&#93;</span> <span class="error">&#91;3 3&#93;</span> <span class="error">&#91;1 1&#93;</span> <span class="error">&#91;-1 -1&#93;</span> <span class="error">&#91;5 5&#93;</span>], :port 6701}]<br/>
    ```</li>
</ul>


<p>    Now three workers are running.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    Thanks @HeartSaVioR. It looks good.</p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    I'd like to get reviewed by @kishorvpatil if (but only if) he's available since it removes the changeset of <a href="https://issues.apache.org/jira/browse/STORM-1561" title="Supervisor should relaunch worker if assignments have changed" class="issue-link" data-issue-key="STORM-1561"><del>STORM-1561</del></a>. Since it's critical issue and related to race condition I'd also like to get reviewed by more contributors / committers. <br/>
    And that would be really nice if affected users can apply this patch to see if this helps. I'll ping reporters for relevant issues and ask them to help testing.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user satishd commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @HeartSaVioR added resolution details in the description of the PR. Adding one more point here: `(kill-existing-workers-with-change-in-components supervisor existing-assignment new-assignment)` is not needed in `sync-supervisor` as it kills all the workers with invalid state which includes the old workers and rebalanced workers.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user satishd commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    +1 LGTM. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    Added some comments. I'll squash once reviewing is done.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user satishd commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @HeartSaVioR You may want to use `(declare rm-topo-files)` in the start so that you do not need to move `defn sync-processes` for handling forward references. This may make it easy to review the code. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @satishd Yeah it would be really better. Thanks for the tip. Applied.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528#discussion_r69152524" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528#discussion_r69152524</a></p>

<p>    &#8212; Diff: storm-core/src/clj/org/apache/storm/daemon/supervisor.clj &#8212;<br/>
    @@ -425,6 +433,13 @@<br/>
      ". State: " state<br/>
      ", Heartbeat: " (pr-str heartbeat))<br/>
     (shutdown-worker supervisor id)))<br/>
    +<br/>
    +    (doseq <span class="error">&#91;storm-id all-downloaded-storm-ids&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    This looks like you removed the check where we used to keep the jars even the assignement goes away. We had another issue where if the topology is rebalanced and the assignment goes away <br/>
    and immediately comes back to the same supervisor , it tries re-download the jar and if it happens to be a large jar file than there are chances that supervisor worker gets killed before it even starts as the jar download takes longer time.<br/>
    IMO , i don't see a reason to remove the storm jars on a rebalance or if the assignement goes away. They should be removed when the topology gets killed.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @HeartSaVioR  I had a fix before for similar issue here <a href="https://github.com/apache/storm/pull/401/files" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/401/files</a> . I don't see much of the code here. Can you take a look if that particular case is handled here.</p>

<p>    @revans2 @kishorvpatil @d2r can you also take a look at this PR.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @harshach <br/>
    `sync-supervisor` writes local assignment after code download for topology is completed. <br/>
    Since `sync-processes` launches worker based on local assignment, there's a guarantee that code download is completed before launching worker.</p>

<p>    Btw, as you see the diff, <a href="https://issues.apache.org/jira/browse/STORM-130" title="[Storm 0.8.2]: java.io.FileNotFoundException: File &#39;../stormconf.ser&#39; does not exist" class="issue-link" data-issue-key="STORM-130"><del>STORM-130</del></a> is not on recent code. Not sure why it's removed.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528#discussion_r69245914" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528#discussion_r69245914</a></p>

<p>    &#8212; Diff: storm-core/src/clj/org/apache/storm/daemon/supervisor.clj &#8212;<br/>
    @@ -425,6 +433,13 @@<br/>
      ". State: " state<br/>
      ", Heartbeat: " (pr-str heartbeat))<br/>
     (shutdown-worker supervisor id)))<br/>
    +<br/>
    +    (doseq <span class="error">&#91;storm-id all-downloaded-storm-ids&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    @harshach <br/>
    I agree that it would be nice to keep topology code until topology gets killed, but since this is same to what it was, I guess we can handle this as new issue for improvement.<br/>
    Btw, `assignment-snapshot` seems hiding no topology info (I mean killed) vs no assignment. We should let assignment-snapshot distinguish both.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @harshach <br/>
    And sync-processes (`get-valid-new-worker-ids`) checks topology files before launching workers.<br/>
    If there's an issue which corrupts topology code or not having topology code, sync-processes doesn't launch worker, and next runs of sync-supervisor will try to download topology code again.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528#discussion_r69497481" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528#discussion_r69497481</a></p>

<p>    &#8212; Diff: storm-core/src/clj/org/apache/storm/daemon/supervisor.clj &#8212;<br/>
    @@ -425,6 +433,13 @@<br/>
      ". State: " state<br/>
      ", Heartbeat: " (pr-str heartbeat))<br/>
     (shutdown-worker supervisor id)))<br/>
    +<br/>
    +    (doseq <span class="error">&#91;storm-id all-downloaded-storm-ids&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    @harshach Would I file a new issue regarding your suggestion, or do you want to file it?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @harshach I commented your concern so could you revisit this?<br/>
    @ptgoetz Could you also take a look?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    @HeartSaVioR I am +1 .</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528#discussion_r70016445" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528#discussion_r70016445</a></p>

<p>    &#8212; Diff: storm-core/src/clj/org/apache/storm/daemon/supervisor.clj &#8212;<br/>
    @@ -425,6 +433,13 @@<br/>
      ". State: " state<br/>
      ", Heartbeat: " (pr-str heartbeat))<br/>
     (shutdown-worker supervisor id)))<br/>
    +<br/>
    +    (doseq <span class="error">&#91;storm-id all-downloaded-storm-ids&#93;</span><br/>
    &#8212; End diff &#8211;</p>

<p>    Never mind. Filed <a href="https://issues.apache.org/jira/browse/STORM-1952" title="Keeping topology code for supervisor until topology got killed" class="issue-link" data-issue-key="STORM-1952"><del>STORM-1952</del></a>(<a href="https://issues.apache.org/jira/browse/STORM-1952" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-1952</a>)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1529" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1529</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1528</a></p>

<p>    Thanks @HeartSaVioR . Merged into 1.0.x, 1-x, master branches.</p>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1933">STORM-1933</a>]</i>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1879">STORM-1879</a>]</i>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1915">STORM-1915</a>]</i>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-2076">STORM-2076</a>]</i>

**Comment by jira-importer** [comment]
<i>[Epic: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1855%20label%3Aepic">STORM-1855</a>]</i>
